src/hffs/fs.py

```diff
@@ -118,35 +89,76 @@ def _resolve_repo_id(self, path: str) -> Tuple[str, str, str]:
+        revision = None
         if path.count("/") > 0:
-            repo_id_with_namespace = "/".join(path.split("/")[:2])
-            path_in_repo_with_namespace = "/".join(path.split("/")[2:])
-            repo_id_without_namespace = path.split("/")[0]
-            path_in_repo_without_namespace = "/".join(path.split("/")[1:])
-            if self._repo_exists(repo_id_with_namespace, repo_type=repo_type):
-                repo_id = repo_id_with_namespace
-                path_in_repo = path_in_repo_with_namespace
-            elif self._repo_exists(repo_id_without_namespace, repo_type=repo_type):
-                repo_id = repo_id_without_namespace
-                path_in_repo = path_in_repo_without_namespace
+            if "@" in path:
+                repo_id, revision = path.split("@", 1)
+                if "/" in revision:
+                    revision, path_in_repo = revision.split("/", 1)
+                else:
+                    path_in_repo = ""
+                revision = unquote(revision)
+                if not self._repo_exists(repo_id, repo_type=repo_type):
+                    raise FileNotFoundError(f"No such repository: '{repo_id}'")
+                return repo_type, repo_id, revision, path_in_repo
             else:
-                raise FileNotFoundError(f"No such repository: '{repo_id_with_namespace}'")
+                revision = None
+                repo_id_with_namespace = "/".join(path.split("/")[:2])
+                path_in_repo_with_namespace = "/".join(path.split("/")[2:])
+                repo_id_without_namespace = path.split("/")[0]
+                path_in_repo_without_namespace = "/".join(path.split("/")[1:])
+                if self._repo_exists(repo_id_with_namespace, repo_type=repo_type):
+                    repo_id = repo_id_with_namespace
+                    path_in_repo = path_in_repo_with_namespace
+                elif self._repo_exists(repo_id_without_namespace, repo_type=repo_type):
+                    repo_id = repo_id_without_namespace
+                    path_in_repo = path_in_repo_without_namespace
+                else:
+                    raise FileNotFoundError(f"No such repository: '{repo_id_with_namespace}'")
         else:
+            if "@" in path:
+                path, revision = path.split("@", 1)
+                revision = unquote(revision)
+            else:
+                revision = None
             if self._repo_exists(path, repo_type=repo_type):
                 repo_id = path
                 path_in_repo = ""
             else:
                 # can't list repositories at the namespace level
                 raise NotImplementedError("Acces to repositories lists is not implemented.")
 
-        return repo_type, repo_id, path_in_repo
+        return repo_type, repo_id, revision, path_in_repo
+
+    def _resolve_path_with_revision(self, path: str, revision: Optional[str]) -> Tuple[str, str, str, str]:
```

Comment:
Returned revision can still be `None` (or should return `huggingface_hub.constants.DEFAULT_REVISION` otherwise)

```suggestion
    def _resolve_path_with_revision(self, path: str, revision: Optional[str]) -> Tuple[str, str, Optional[str], str]:
```

Reply:
Thinking twice about it maybe we should return a `str` revision all the time, defaulting to `huggingface_hub.constants.DEFAULT_REVISION` if not set. This logic can be implemented in `resolve_path`.

This would be coherent with the fact that we return `model` as `repo_type` when not defined. And having a revision always set can avoid mistakes in the future.